strawberry/types/type_resolver.py

```diff
@@ -222,60 +222,181 @@ def _resolve_types(fields: List[FieldDefinition]) -> List[FieldDefinition]:
 
 
 def _get_fields(cls: Type) -> List[FieldDefinition]:
-    fields = []
+    """Get all the strawberry field definitions off a strawberry.type cls
+
+    A strawberry.type class can have 3 different field types:
+        1.  A pure dataclass field, defined without any use of the
+            strawberry.field function/decorator. Will not have a
+            field_definition --> need to create one.
+        2a. A strawberry.field that has not received a resolver. Will be a
+            stand-alone FieldDefinition object.
+        2b. A resolver modified with strawberry.field(resolver=...) or decorated
+            with @strawberry.field. Function object will have a monkey-patched
+            _field_definition attribute.
+
+        Type #2a will have a corresponding Type #1 item. Items of Type #2b will
+        also have a corresponding item if defined with the strawberry.field
+        function (i.e. not the decorator).
+
+        This function needs to return a list of FieldDefinitions, one for each
+        field item, without duplicates, while also paying attention the `name`
+        field.
+
+        Final `name` attribute priority:
+        1. Type #2 `name` attribute. This will be defined with an explicit
+           strawberry.field(name=...). No camelcase-ification will be done,
+           as the user has explicitly stated the field name.
+        2. field name on cls. Will exist for both Type #1 and Type #2 fields.
+
+           Field names will be converted to camelcase
+
+           If a field is Type #2 field, but never received a resolver, there is
+           no way to determine the typing of the field, and it is invalid.
+
+    """
+    field_definitions: Dict[str, FieldDefinition] = {}
+
+    # Type #1 fields
+    type_1_fields: Dict[str, dataclasses.Field] = {
+        field.name: field for field in dataclasses.fields(cls)
+    }
+
+    # Type #2 fields
+    type_2_fields = {}
+    for field_name, field in cls.__dict__.items():
+        if hasattr(field, "_field_definition"):
+            type_2_fields[field_name] = field
+
+    for field_name, field in type_2_fields.items():
 
-    # get all the fields from the dataclass
-    dataclass_fields = dataclasses.fields(cls)
+        # Check if there is a matching Type #1 field:
+        if field_name in type_1_fields:
+            # Stop tracking the Type #1 field, an explict strawberry.field was
+            # defined
+            type_1_fields.pop(field_name)
+        # Otherwise, ensure that a resolver has been specified
+        else:
+            field_definition: FieldDefinition = field._field_definition
+            if field_definition.base_resolver is None:
+                raise WindowsError(...)
```

Comment:
This actually doesn't get hit. An exception (a TypeError with reasonable detail) is thrown in `type_resolver._get_fields` when processing the dataclass. 

Should we just use that exception, or do we want to convert it to a Strawberry exception?

Reply:
(note that WindowsError is a fill-in, and won't be in the final commit)

Reply:
Issue already tracked in #396 . Will address in future PR